lib/postgresql/native_postgres.nit

```diff
@@ -0,0 +1,104 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module native_postgres is pkgconfig("libpq")
+
+in "C header" `{
+  #include <libpq-fe.h>
+`}
+
+extern class ExecStatusType `{int`}
+  new empty           `{ return PGRES_EMPTY_QUERY; `}
+  new command_ok      `{ return PGRES_COMMAND_OK; `}
+  new tuples_ok       `{ return PGRES_TUPLES_OK; `}
+  new copy_out        `{ return PGRES_COPY_OUT; `}
+  new copy_in         `{ return PGRES_COPY_IN; `}
+  new bad_response    `{ return PGRES_BAD_RESPONSE; `}
+  new nonfatal_error  `{ return PGRES_NONFATAL_ERROR; `}
+  new fatal_error     `{ return PGRES_FATAL_ERROR; `}
+
+  fun is_ok: Bool `{return self == PGRES_TUPLES_OK || self == PGRES_COMMAND_OK; `}
+
+  redef fun to_s import NativeString.to_s `{
+    char * err = PQresStatus(self);
+    if(err == NULL) err = "";
+    return NativeString_to_s(err);
+  `}
+end
+
+extern class ConnStatusType `{int`}
+  new connection_ok `{ return CONNECTION_OK; `}
+  new connection_bad `{ return CONNECTION_BAD; `}
+
+  fun is_ok: Bool `{return self == CONNECTION_OK; `}
+end
+
+
+extern class NativePostgres `{PGconn *`}
+
+  # Connect to a new database using the conninfo string as a parameter
+  new pq_connectdb(conninfo: String) import String.to_cstring `{
+    PGconn * self = NULL;
+    self = PQconnectdb(String_to_cstring(conninfo));
+    return self;
+  `}
+
+  # Submits a query to the server and waits for the result
+  # returns the ExecStatustype of the query
+  fun pq_exec(query: String):ExecStatusType import String.to_cstring `{
+    PGresult *res = PQexec(self, String_to_cstring(query));
+    return PQresultStatus(res);
+  `}
+
+  # Prepares a statement with the given parameters
+  fun pq_prepare(stmt: String, query: String, nParams: Int):ExecStatusType import String.to_cstring `{
+    const char * stmtName = String_to_cstring(stmt);
+    const char * queryStr = String_to_cstring(query);
+    PGresult * res = PQprepare(self, stmtName, queryStr, nParams, NULL);
+    return PQresultStatus(res);
+  `}
+
+  fun pq_exec_prepared(stmt: String, nParams: Int, values: Array[String], pLengths: Array[Int], pFormats: Array[Int], resultFormat: Int):ExecStatusType import String.to_cstring, Array[String].[] `{
+    const char * stmtName = String_to_cstring(stmt);
+    const char * paramValues[nParams];
+    for(int i = 0; i < nParams; i++)
+      paramValues[i] = String_to_cstring(Array_of_String__index(values, i));
+    PGresult * res = PQexecPrepared(self, stmtName, nParams, paramValues, (const int*) pLengths, (const int*) pFormats, resultFormat);
```

Comment:
The part of `pLengths` and `pFormats` should not work... `Array[Int]` is a complex Nit type, in C it is represented by an opaque pointer, it can't be cast to an `int*`.

You could either create a C array as you did with `paramValues` and for `pLengths` it could be completely computed on the C side.

Honestly, it's not an easy method to wrap for Nit...